quad.go

```diff
@@ -0,0 +1,159 @@
+// the weights and locations specified by inter. That is, Fixed estimates
+//  int_min^max f(x) dx ≈ \sum_i w_i f(x_i)
+// If inter == nil, an acceptable default is chosen. If inter != nil, it is
+// assumed that the properties of the integral match the assumptions of inter.
+// For example, Legendre assumes that the integration bounds are finite. If
+// inter is also a FixedLocationsSingler, the quadrature points are computed
+// individually rather than as a unit.
+//
+// If concurrent <= 0, f is evaluated serially, while if concurrent > 0 f is
+// evaluated may be evaluated with at most concurrent simultaneous evaluations.
+func Fixed(f func(float64) float64, min, max float64, n int, inter FixedLocationer, concurrent int) float64 {
+	// TODO(btracey): When there are Hermite polynomial quadrature, add an additional
+	// example to the documentation comment that talks about weight functions.
+	if n <= 0 {
+		panic("quad: non-positive number of locations")
+	}
+	if min >= max {
+		panic("quad: min >= max")
+	}
+	if math.IsInf(min, 1) {
+		panic("quad: min = inf")
+	}
+	if math.IsInf(max, -1) {
+		panic("quad: max = -inf")
+	}
+	intfunc := f
+	// If inter is non-nil it is assumed that the function and the constraints
+	// of inter are aligned. If is is nil, wrap the function and do something
+	// reasonable.
+	// TODO(btracey): Replace wrapping with other quadrature rules when
+	// we have rules that support indefinite integrals.
+	if inter == nil {
+		// int_a^b f(x)dx = int_u^-1(a)^u^-1(b) f(u(t))u'(t)dt
+		switch {
+		case math.IsInf(max, 1) && math.IsInf(min, -1):
+			// u(t) = (t/(1-t^2))
+			min = -1
+			max = 1
+			intfunc = func(x float64) float64 {
+				v := 1 - x*x
+				return f(x/v) * (1 + x*x) / (v * v)
+			}
+		case math.IsInf(max, 1):
+			// u(t) = a + t / (1-t)
+			a := min
+			min = 0
+			max = 1
+			intfunc = func(x float64) float64 {
+				v := 1 - x
+				return f(a+x/v) / (v * v)
+			}
+		case math.IsInf(min, -1):
+			// u(t) = a - (1-t)/t
+			a := max
+			min = 0
+			max = 1
+			intfunc = func(x float64) float64 {
+				return f(a-(1-x)/x) / (x * x)
+			}
+		}
+		inter = Legendre{}
+	}
+	singler, isSingler := inter.(FixedLocationSingler)
+
+	var xs, weights []float64
+	if !isSingler {
+		xs = make([]float64, n)
+		weights = make([]float64, n)
+		inter.FixedLocations(xs, weights, min, max)
+	}
+
+	if concurrent > n {
+		concurrent = n
+	}
+
+	if concurrent <= 0 {
+		// Evaluate in serial.
+		if isSingler {
+			var integral float64
+			for k := 0; k < n; k++ {
+				x, weight := singler.FixedLocationSingle(n, k, min, max)
+				integral += weight * intfunc(x)
+			}
+			return integral
+		}
+		var intergal float64
+		for i, x := range xs {
+			intergal += weights[i] * intfunc(x)
+		}
+		return intergal
+	}
+
+	// Evaluate concurrently
+	tasks := make(chan int)
+
+	// Launch distributor
+	go func(tasks chan<- int) {
+		for i := 0; i < n; i++ {
+			tasks <- i
+		}
+		close(tasks)
+	}(tasks)
```

Comment:
I seem to remember you telling me in another PR that it was standard to pass channels as input parameters to help with the directionality. If I'm remembering incorrectly, happy to remove.